### ISSUE TITLE
Predict memory leak

### ISSUE THREAD

**Issue by lgvaz** [airctic/icevision#692]
## 🐛 Bug
The prediction contains a reference to the image array. If any stage of the workflow starts accumulating the predictions objects (as it currently happens in `COCOMetric` this will result in a memory leak.

`COCOMetric` currently uses a "quickfix" to make sure the images are not being stored (take a look at `accumulate`), but we need a more general solution.

**Comment by rsomani95** [comment]
Based on our joing session the otherday on the machine with 16GB RAM, it looks like there's memory leaks happening elsewhere too, right?

As a refresher -- we'd turned off `COCOMetrics` and the process kept crashing faster proportial with `num_workers` in the `DataLoader`

**Comment by lgvaz** [comment]
Yeah, we need to investigate this deeper. Let's have a call where we use a profiling tool like described [here](https://medium.com/zendesk-engineering/hunting-for-memory-leaks-in-python-applications-6824d0518774) to find the issue.

**Comment by potipot** [comment]
check if is solved by https://github.com/airctic/icevision/pull/737

**Comment by potipot** [comment]
Do you know if this is specific for pytorch lightning or fastai? I'm facing this issue too and investigating it RN

**Comment by potipot** [comment]
I found that the problem comes from record loading. In fact, the memory use grows only in the 0th epoch and then keeps constant. This led me to thinking that the Dataset is loaded into memory as a whole and this is indeed what happens.

When calling `Dataset`'s getitem here:
```python
    def __getitem__(self, i):
        record = self.records[i].load()
        if self.tfm is not None:
            record = self.tfm(record)
        return record
```
we change the state of records, which are a field of `Dataset` class and they persist in memory as long as the Dataset and Dataloader. 
When I added a manual record unloading at the end of training loop, the memory build-up vanishes:
```python
    def training_step(self, batch, batch_idx):
        (xb, yb), records = batch
        preds = self(xb, yb)

        loss = efficientdet.loss_fn(preds, yb)

        for k, v in preds.items():
            self.log(f"train/{k}", v)
        [record.unload() for record in records]
        return loss
```

Perhaps we should have a custom Data fetcher that takes care of the cleanup at the end of training loop? Or maybe this is already implemented but operates on a deepcopy of the record rather than the object bound with `Dataset` and `Dataloader` ?

**Comment by lgvaz** [comment]
Interesting question: if we use coco metric, does it solve this issue? Because coco metric does call `unload` hahaha

**Comment by potipot** [comment]
Yeah, my previous comment is already outdated :D I'm investigating this still. It's true using COCOMetrics reduces memory use, so as you said there must be some references hanging. 

As far as I understand now, when using num_workers>0 each spawned process loads the record and returns the deepcopy of it. The img data doesn't persist in Dataset but is hanging somewhere else. It is the worker processed that hold the memory. 

Here are some tests that I run:
1. using cocometric and pytorch lightning:
![memory-profile](https://user-images.githubusercontent.com/52150545/112990153-05818c80-9166-11eb-88a1-0d3a743c32a5.png)
2. using cocometric and fastai: ![memory-profile-fastai](https://user-images.githubusercontent.com/52150545/112990276-29dd6900-9166-11eb-885a-1ae8639e25bb.png)
3. not using cocometric and pytorch lightning: 
![memory-profile-nometric](https://user-images.githubusercontent.com/52150545/112990678-9c4e4900-9166-11eb-8cb1-cac2ed7689a2.png)

**Comment by potipot** [comment]
might be linked with https://github.com/pytorch/pytorch/issues/13246 again, sorry Lucas :D

**Comment by potipot** [comment]
a temoporary solution that seems to reduce memory leakage is to use num_workers not larger than the number of physical cores of the cpu as described here https://discuss.pytorch.org/t/num-workers-in-dataloader-will-increase-memory-usage/28522/9

**Comment by boji123** [comment]
after worker extract a batch, memory will not release automatically, memory will release only after one epoch finish
try dumping batch data to disk and save a path in memory, only load it after worker extract a batch and call collate function, by this way the memory will increase quite slower than before

**Comment by lgvaz** [comment]
@boji123 I'm not sure I follow, are you proposing that we pre-save batches to disk?

**Comment by potipot** [comment]
There is one other issue with predict, the tensor_to_image stores the data in floating point resulting in excessive use of memory. Perhaps we could convert the images to uint at this stage? Or would that result in data loss? @lgvaz

**Comment by lgvaz** [comment]
@potipot I don't think it will result in significant data loss, we should indeed do that. You up for sketching a quick PR? =)

**Comment by potipot** [comment]
This fix might be breaking though. For anyone that is using inference and denormalization function, they assume that predictions come in this non-normalized floating point tensor format. 
Probably this is not just a single fix but has impact on the entire inference workflow.

**Comment by lgvaz** [comment]
Hm... Would this mean that we have to denorm inside `convert_raw_predictions`?

**Comment by drscotthawley** [comment]
Hi would someone mind sharing the status on this or proposed workarounds?   I am suddenly stuck, getting GPU memory overruns on all machines of various VRAM sizes (e.g. Colab High RAM, Titan X,...) whenever I try to predict now  -- even though my notebooks were working fine yesterday.  Was using v0.8.1.  Just upgraded via pip to v0.9.0a1, but problem persists.

**Comment by drscotthawley** [comment]
Deleting my earlier comment as I have no workaround.  All my codes now produce memory overrun at prediction.  Decreasing num_workers to as low as 1 did not solve the issue. I do not have this issue with other PyTorch-based libraries, only IceVision. This problem arises even for relatively small datasets on GPUs with VRAMs of 10, 12, and 15GB. 
  I tried rewriting the suggested parser ('ChessParser') to use something other than lists (e.g. fastai's "L") as per a suggestion in PyTorch issue https://github.com/pytorch/pytorch/issues/13246 but was getting "unhashable type" errors -- seems other parts of IceVision demand lists.
Would love any further suggestions on this.

**Comment by drscotthawley** [comment]
Seems the predictor may need a lower batch size than the trainer!  
Restarting the kernel, loading saved weights, and then _dropping the batch size by a factor of 2 for prediction only_ allowed my run to go to completion.

**Comment by lgvaz** [comment]
@drscotthawley 

- Can you post the snippet of code you're using for doing the predictions?
- What is the prediction dataset size? 
- Does this still happens if you set `keep_images=False`?

**Comment by igvas** [comment]
Hello Scott.

Wrong address.

**Comment by FraPochetti** [comment]
This has been resolved.
